Connection move

tonimueller opened the pull request:
This is a simple approach to connect a system with a vicinity search. 

Maybe there could be some more test for the feature, but it should still work.

MartinWenge committed:
adding the template type of Feature Atrributes in FeatureExcludedVolumeSc.h MISSING: all other festures

tonimueller committed:
Generalize FeatureAttributes to take any class as an attribute tag.
This induces changes in all inherit classes of AddBaseClass
and some minor changes in the features and test using FeatureAttributeTag.

tonimueller committed:
Add first test for a new attribute class, tests and class are not finished!

tonimueller committed:
Add a class MonomerLabel as an example for another attribute predicate and write test for it

tonimueller committed:
adapt FeatureChangeAttributes to the new template parameter of FeatureAttributes

tonimueller committed:
start writing base class and specialized class for a connection move

Commit:
add basic move and

Commit:
add a feature to check the connection process

tonimueller committed:
start to write a lattice class for general use, motivated by the FeatureConnectionSc.

tonimueller committed:
finished the implementation for a  general Lattice class and write a small test

tonimueller committed:
finish implmentaion of the connection moves

tonimueller committed:
implement and test a connection Feature which checks if the monomers are allowed to connect with each other

tonimueller committed:
add output streams for the bfm file

Commit:
add an example updater for a connection process and a tracker if the history of connection is important

tonimueller committed:
solve some first issues

tonimueller committed:
adding the FeatureExcludedVolume causes problems if it is excplicitly included in the main program with another template parameter

tonimueller committed:
missing connection in the apply function

MartinWenge committed:
remove some unneded newlines from TestFeatureWall.cpp

Bondoki committed:
* corrections and testing of new feature
    -> renaming of function IsReactive() -> isReactive()
    -> avoiding overloading of << in MonomerReactivity due to ambiguity in file writing

Bondoki commented on include/LeMonADE/feature/FeatureConnectionSc.h:
```diff
@@ -76,6 +96,76 @@ class MonomerReactivity
+  * @param stream output-stream
+  * @param label object of class MonomerReactivity
+  * @return output-stream
+  **/
```
* wrong file output due to ambiguity of "<<" operator in MonomerReactivity

tonimueller replied:
Add a todo in the class description.

Bondoki commented on include/LeMonADE/feature/FeatureConnectionSc.h:
```diff
@@ -56,12 +59,29 @@ class MonomerReactivity
+	{
+	  reactivity = react.IsReactive();
+	  nMaxBonds = react.getNMaxBonds();
+	}
```
* wrong implementation of comparison operator

Bondoki commented on include/LeMonADE/feature/FeatureConnectionSc.h:
```diff
@@ -328,4 +455,150 @@ void FeatureConnectionSc::fillLattice(IngredientsType& ingredients)
+    if(molecules[n].getReactivity()!=reactivity)
+    {
+      strm<<startIndex+1<<"-"<<n<<":"<<reactivity<<std::endl;
+      reactivity=molecules[n].getReactivity();
```
make the call explicit due to incompatible stream operator

Bondoki commented on include/LeMonADE/feature/FeatureConnectionSc.h:
```diff
@@ -328,4 +455,150 @@ void FeatureConnectionSc::fillLattice(IngredientsType& ingredients)
+  }
+  //write final reactivity
+  strm<<startIndex+1<<"-"<<nMonomers<<":"<<reactivity<<std::endl<<std::endl;
+
```
make the call explicit due to incompatible stream operator

Bondoki commented on include/LeMonADE/feature/FeatureConnectionSc.h:
```diff
@@ -56,12 +59,29 @@ class MonomerReactivity
 	uint32_t getNMaxBonds() const {return nMaxBonds;};
 
+	MonomerReactivity getReactivity(){return *this; }
+	void setReactivity(const MonomerReactivity& react)
```
bad practice to call a member function not by its name -> rename to getMonomerReactivity() and setMonomerReactivity()

Bondoki commented on include/LeMonADE/feature/FeatureConnectionSc.h:
```diff
@@ -0,0 +1,621 @@
+	if ( connectionLattice.getLatticeEntry(Pos+dir) == 0 ) return false;
+	//check for maximum number of bonds for the first monomer
+	if ( ingredients.getMolecules()[ID].getNMaxBonds() >= ingredients.getMolecules().getNumLinks(ID) ) return false;
+	uint32_t Neighbor(connectionLattice.getLatticeEntry(Pos+dir)-1);
```
**STRONG LOGICAL ERROR**
should read 
`if ( ingredients.getMolecules().getNumLinks(ID) >= ingredients.getMolecules()[ID].getNMaxBonds() ) return false;`

Bondoki commented on include/LeMonADE/feature/FeatureConnectionSc.h:
```diff
@@ -0,0 +1,621 @@
+	if ( ingredients.getMolecules()[ID].getNMaxBonds() >= ingredients.getMolecules().getNumLinks(ID) ) return false;
+	uint32_t Neighbor(connectionLattice.getLatticeEntry(Pos+dir)-1);
+	//check for maximum number of bonds for the second monomer
+	if ( ingredients.getMolecules()[Neighbor].getNMaxBonds() >= ingredients.getMolecules().getNumLinks(Neighbor) ) return false;
```
**STRONG LOGICAL ERROR**
should read 
`if ( ingredients.getMolecules().getNumLinks(Neighbor) >= ingredients.getMolecules()[Neighbor].getNMaxBonds()  )`

Bondoki commented on include/LeMonADE/updater/UpdaterSimpleConnection.h:
```diff
@@ -0,0 +1,176 @@
+    if ( ingredients.getMolecules()[i].IsReactive() )
+    {
+      uint32_t NLinks(ingredients.getMolecules().getNumLinks(i));
+      uint32_t nIrreversibleBonds;
```
**LOGICAL ERROR** forgotten initialization  
should read 
`uint32_t nIrreversibleBonds = 0`

tonimueller replied:
resolved in 0989b7bf3b432648062dbca93499428926ad7125

Bondoki commented on include/LeMonADE/updater/UpdaterSimpleConnection.h:
```diff
@@ -0,0 +1,176 @@
+  //! number of already reacted monomers
+  uint32_t NReactedSites;
+  //! get conversion of the reaction process
+  double getConversion(){return (1.*NReactedSites)/(1.*NReactiveSites);};
```
Conversion is defined as c= num all established Links / num all maximum Links
Does this calculated a value between [0;1]?

tonimueller replied:
Yes, the value should be between [0:1]. Is there any concern about it?

Bondoki replied:
This return an "undefined" values as the implementation is not correct due to the "counting" of NReactiveSites. Just try this with an arbitrary system.

tonimueller replied:
found the bugs and correct them see a0b7e7c0070cf3f82d6fe0752300617be96b2afb

Bondoki commented on include/LeMonADE/feature/FeatureConnectionSc.h:
```diff
@@ -0,0 +1,621 @@
+  public:
+
+	//! Standard constructor- initially the reactivity is set to false.
+	MonomerReactivity():reactivity(false),nMaxBonds(2){}
```
default value should be used `nMaxBonds(0)`
as this is has no effect if reactivity is false. Anyway this can make things inconsitent as we define a maximum connectivity for all monomers in the Molecules class as template argument `max_connectivity`
There should be a check in the read-in function to test against the "default" maximum conncetivity for consistency reason.

tonimueller replied:
add check, see #cd5263f8c07565625017f03134a10b3a488525bf

Bondoki commented on include/LeMonADE/feature/FeatureConnectionSc.h:
```diff
@@ -0,0 +1,621 @@
+	MonomerReactivity& operator= (const MonomerReactivity source)
+	{
+	  reactivity=source.isReactive();
+	  nMaxBonds=source.getNMaxBonds();
```
There should be a check in the function to test against the "default" maximum connectivity for consistency reason.

tonimueller replied:
see cd5263f8c07565625017f03134a10b3a488525bf

Bondoki committed:
* fix the main LOGICAL ERRORS in FeatureConnectionSc.h
* renaming Bonds -> Links
* renaming 'N*' -> 'num*'

Bondoki commented on include/LeMonADE/utility/AttributeTags.h:
```diff
@@ -0,0 +1,215 @@
+    LabelID=in_[2] ;
+  }
+  //! get the chain ID, number of labels and label id 
+  const std::vector<uint32_t>& getAll() const { 
```
**LOGICAL ERROR** : reference to local variable
should read `const std::vector<uint32_t> getAll() const`

tonimueller replied:
see commit  c5ffb9236092d4eb6744cb8666085bef69c27faf

Bondoki commented on include/LeMonADE/utility/AttributeTags.h:
```diff
@@ -0,0 +1,215 @@
+  * @return input-stream
+  **/
+  std::istream& operator >> (std::istream& stream, MonomerLabel & label)
+  {
```
`ERROR HANDLING?`

tonimueller replied:
see commit  c5ffb9236092d4eb6744cb8666085bef69c27faf

Bondoki commented on include/LeMonADE/utility/AttributeTags.h:
```diff
@@ -0,0 +1,215 @@
+  }
+  
+private:
+  uint32_t ChainID;
```
* explain variables and comment
* use lower case style and meaningful naming

tonimueller replied:
see commit  c5ffb9236092d4eb6744cb8666085bef69c27faf

Bondoki commented on include/LeMonADE/updater/UpdaterSimpleConnection.h:
```diff
@@ -0,0 +1,176 @@
+	}
+	uint32_t nTags(TagedMonomers.size());
+	
+	for(uint32_t i =0; i < nTags;i++)
```
This implicit increase the MCS. Maybe an alternative approach is useful? See latest commit.

tonimueller replied:
Do you mean that the two moves implicitly increase the MCS? This is resolved in 1786028e4222aab18b7692c9b395a45d8e46506a.

Bondoki commented on include/LeMonADE/feature/FeatureConnectionSc.h:
```diff
@@ -0,0 +1,621 @@
+	if ( ingredients.getMolecules()[Neighbor].getNMaxBonds() >= ingredients.getMolecules().getNumLinks(Neighbor) ) return false;
+	//check if the two monomers are already connected
+	if ( ingredients.getMolecules().areConnected(ID,Neighbor) ) return false;
+	//if still here, then the two monomers are allowed to connect 
```
**LOGICAL ERROR**
if one has non-periodic boundary conditions, there's no check forbidding bonds over box-size

tonimueller replied:
This is partially resolved in 1786028e4222aab18b7692c9b395a45d8e46506a by adding the move into the FeatureBox.h. 
Still there are problems with the connecting images where only the refolded bondvector fit into the bondvector set but the absolute vector doesnt fit!

tonimueller replied:
Allow only connection if the bonds are in the bondvector set ca6c24b67f63bfe7676cf0652ab1d6778ee6898f .
This is the most relevant case for most applications, but must be changed for creating connection between images of monomer.

Bondoki commented on include/LeMonADE/feature/FeatureConnectionSc.h:
```diff
@@ -0,0 +1,621 @@
+	typedef LOKI_TYPELIST_1(MonomerReactivity) monomer_extensions;
+	
+	//! 
+// 	typedef LOKI_TYPELIST_1(FeatureExcludedVolumeSc<>) required_features_back;
```
FeatureExcludedVolumeSc<> should be mandatory as this algorithm implicitly uses excluded volume condition

Bondoki replied:
this feature should also require the "Feature*UnsaveCheck" e.g. FeatureMoleculesIOUnsaveCheck.h, FeatureBondsetUnsaveCheck.
Is there any possibility to "overwrite" the "save" features in the LOKi-List. If not, maybe it's a good practice to implement a test in ConfigureSystem to check for incompatible feature and throw an error.

tonimueller replied:
The problem is that specialized features of FeatureExcludedVolumeSc<> given in the feature list of the main file lead to ambiguous function calls. We certainly need to change this, but for now I would just kick out the required_features_back in this feature and open an issues, see a331711cd3f5e6da9589e7cc3c355606d49071ba .

Bondoki commented on include/LeMonADE/feature/FeatureBox.h:
```diff
@@ -311,8 +311,8 @@ class FeatureBox : public Feature
+	bool checkMove(const IngredientsType& ingredients, const MoveAddMonomerBase<AddMoveType, TagType>& addmove) const
 	{
 		VectorInt3 pos=addmove.getPosition();
 
```
Implement check for MoveConnectSc to deal with periodic boundary conditions

Bondoki committed:
* implement MoveConnectSc to FeatureBox
* alternative implementation in UpdaterSimpleConnection.h
* rename Bonds -> Links
* correct logical error in AttributeTags.h
* correct typo
-> todo: adding test
-> todo: implement checker for incompatible features
-> todo: prove suggestions

----> reject the contribution at this state

Bondoki reviewed (changes requested):
reject pull request #90 
see my commits https://github.com/tonimueller/LeMonADE/commit/1786028e4222aab18b7692c9b395a45d8e46506a

tonimueller committed:
Update AttributeTags.h

I applied the suggested changes by @Bodonki:
- lower case variables
- error handling in the input stream operator  
- some more comments

tonimueller committed:
change the function call for the AttributeTag class MonomerLabel in the tests

tonimueller committed:
add a check in ReadReactivity::execute() for the number of maximum bonds given from the monomer extension to not exceed the number of maximum bonds given by ingredients. Adapt this in the corresponding test.

tonimueller committed:
FeatureConnectionSc.h: connections are only allowed if the bondvector is in the bondvectorset!

tonimueller committed:
kick out the feature_required_back for compatibility reasons

tonimueller reviewed (commented):
Read through the suggestions and add commits to implement some changes.
(Todos of  commit 1786028e4222aab18b7692c9b395a45d8e46506a are not processed )

HeisenbergGuo committed:
Update FeatureConnectionSc.h

supposed be some bugs? pls check and confirm

Commit:
Merge pull request #1 from HeisenbergGuo/patch-1

Update FeatureConnectionSc.h

tonimueller committed:
Add the (overloaded) function for the connection moves into the FeatureBondset*.h

tonimueller committed:
Merge branch 'ConnectionMove' of https://github.com/tonimueller/LeMonADE into ConnectionMove

tonimueller committed:
Bugfix: Comment not corretly set. Erase it because unneeded.

tonimueller committed:
add a getter and setter function to apply the move in the feature FeatureConnectionSc

tonimueller committed:
test the init, check and apply functions as well as the constructors and the setter/getter

tonimueller committed:
add more tests for FeatureConnectionSc and erase bugs. Correct the stream operators in AttributeTags and for the Reactivity

tonimueller committed:
write a test for the getConversion() function of the updater UpdaterSimpleConnection.

tonimueller commented:
At the moment I cant see anything missing, so please tell me if there is any issue where I should take care of.

tonimueller committed:
Correct the checkMove for FeatureBondset to use the isValidStrongCheck. implement a test for the connection of periodic images in the TestFeatureBondsetUnsaveCheck.cpp

tonimueller committed:
Fix the stream operator in TrackConnection: Still not sure if the molecules stream operator gets overwritten.

tonimueller committed:
Test the WriteReactivity class. There was a problem that the output stream operator of the reactiviy class still overwrittes the output for the coordinates. Thus I changed back the WriteReactivity to an explicit writting. I dont like that -_-

tonimueller committed:
fix: forgot to erase a statement.

tonimueller committed:
Closes #95
Add again the output stream operator for the MonomerReactivity.

tonimueller committed:
Add checkMove for MoveLocalScDiag and revise TrackConnectio class

MartinWenge commented:
I am fine with this version now let's put it in the development branch and do further fine tuning there. This pull request is way too large anyway ;)
